### ISSUE TITLE
Two sACN devices on one universe

### ISSUE THREAD

**Issue by AlexWHughes** [PennazSoftware/homebridge-dmxlight-plugin#2]
**Describe The Bug:**

**To Reproduce:**
Adding two DMX fixtures using the same universe seems to cause issues when you control only one of them. Add in two "elements" to reproduce

**Expected behavior:**
When I control the first light the second one shouldn't turn off. It seems like when I am controlling one it sends zero values to the other light. 

**Logs:**

```
Show the Homebridge logs here, remove any sensitive information.
```

**Plugin Config:**

{
    "name": "lights",
    "ipAddress": "xxx.xxx.xxx.xxx",
    "platform": "DMXLightHomebridgePlugin",
    "accessories": [
        {
            "name": "RGB-1",
            "id": "lx-1",
            "driverName": "sacn",
            "dmxStartChannel": 1,
            "dmxChannelCount": 192,
            "dmxUniverse": 1,
            "colorOrder": "rgb"
        },
		     "name": "RGB-2",
            "id": "lx-2",
            "driverName": "sacn",
            "dmxStartChannel": 193,
            "dmxChannelCount": 4,
            "dmxUniverse": 1,
            "colorOrder": "rgb"
		{
    ]
}

**Comment by AlexWHughes** [comment]
No problem. It's a minor thing but would be great if there is an easy fix. I suspect its sending a full universe of data every time rather than just the affected channels. 

Would also be great if it worked with RGBW.

**Comment by AlexWHughes** [comment]
Any luck? Is it a complex fix?

**Comment by AlexWHughes** [comment]
Hello, Did you ever work out a solve for this?

**Comment by PennazSoftware** [comment]
I haven’t. All of my DMX lights are at my house in Seattle and I’ve been living at my Palm Springs place since December and haven’t had a chance to test a fix.  The good thing is that I’ll be home in two weeks and will look into it. Sorry it’s taken so long.

-Dan

On May 1, 2022, at 10:43 PM, AlexWHughes ***@***.***> wrote:

﻿

Hello, Did you ever work out a solve for this?

**Comment by AlexWHughes** [comment]
No problem. Thanks!

**Comment by AlexWHughes** [comment]
Any luck? Let me know if there is anything I can do to help?

**Comment by AlexWHughes** [comment]
Hello, 

Is there update to this issue?

**Comment by AlexWHughes** [comment]
I'm happy to put some $$$ behind this one if that helps.

**Comment by PennazSoftware** [comment]
I’ll have another look at getting it to work but my previous attempt pointed to a limitation with a dmx plugin I have a dependency on and I wasn’t able to figure out a workaround.

I’m out of town right now but will have a look when I return on the 26th.

-Dan

On Dec 18, 2022, at 2:32 AM, AlexWHughes ***@***.***> wrote:

﻿

I'm happy to put some $$$ behind this one if that helps.

**Comment by AlexWHughes** [comment]
As we discussed nearly a year ago (sorry about the harassment) I believe the issue is the module is sending a full universe of data so when you try to control one light it sends data to both. If there isn't an easy fix then we can close this out and move on as its current state should be fine for most people.

**Comment by alvarezb** [comment]
I ran into the same issue, and did some digging. The sACN protocol [as defined in this spec document](https://tsp.esta.org/tsp/documents/docs/ANSI_E1-31-2018.pdf) only seems to support sending an array of DMX values that start at address 0, and the [e131 library this project is using](https://github.com/hhromic/e131-node/blob/master/lib/e131/dmp_layer.js) will fill in 0 values for any channels before the ones we are trying to update.

This means that turning on your "RGB-2" light and *then* turning on your "RGB-1" should work, but if you do it in the reverse order, the 0 "padding" values on addresses 0-192 sent by controlling RGB-2 will turn off all the channels used for RGB-1.

Given this is an issue for me as well, I'll investigate whether I can cache the state of the whole universe to send the "last used" values for each channel every time an update is sent, instead of sending mostly 0s that turn off other lights at lower addresses.

**Comment by PennazSoftware** [comment]
I moved the code outside of HomeBridge and wrote a simple test application.  I have a E131 controller (Holiday Coro Alphapix 4) configured with 2 universes (300 lights on each).  The application issues the following commands with a 1-second delay between them:

1. Universe 1 Off
2. Universe 2 Off
3. Universe 1 White
4. Universe 2 White
5. Universe 1 Red
6. Universe 2 Green
7. Universe 1 Blue
8. Universe 2 Purple
9. Universe 2 Off

The application controls the lights correctly...issuing a command on one universe does not impact the other universe.  I also tried instantiating a new instance of my dmx library midway through the procedure and it still ran correctly.

I suspect at this point the problem stems from running within HomeBridge and am going to conduct some testing in that environment.

**Comment by PennazSoftware** [comment]
Update: I've made good progress. I now have a clear repro outside of HomeBridge. This is in fact the problem that alvarezb described and has to do with me resetting the buffer each time a new call is made. I am starting to work on a solution that maintains the state of the buffer between calls and only updates that portion of the buffer that needs to be. 

Stay tuned!

**Comment by AlexWHughes** [comment]
Amazing! thank you so much

**Comment by PennazSoftware** [comment]
@AlexWHughes , I have published a new version (1.1.16) which should address the problem you are having. I will leave this issue open until I hear back from you. Please let me know if you see any problems with this fix.

**Comment by AlexWHughes** [comment]
Looks like this solved it. Thank you so much!
